src/machine/machine_atmega.go

```diff
@@ -155,3 +155,94 @@ func (uart UART) WriteByte(c byte) error {
 	avr.UDR0.Set(c) // send char
 	return nil
 }
+
+// SPIConfig is used to store config info for SPI.
+type SPIConfig struct {
+	Frequency uint32
+	SCK       Pin
+	SDO       Pin
+	SDI       Pin
+	LSBFirst  bool
+	Mode      uint8
+}
+
+// SPI is for the Serial Peripheral Interface
+// Data is taken from http://ww1.microchip.com/downloads/en/DeviceDoc/ATmega48A-PA-88A-PA-168A-PA-328-P-DS-DS40002061A.pdf page 169 and following
+type SPI struct{}
+
+// Configure is intended to setup the SPI interface.
+func (spi SPI) Configure(config SPIConfig) error {
+
+	// Make the defaults meaningful
+	if config.Frequency == 0 {
+		config.Frequency = 4000000
+	}
+	if config.SCK == 0 {
+		config.SCK = SPI0_SCK_PIN
+	}
+	if config.SDO == 0 {
+		config.SDO = SPI0_SDO_PIN
+	}
+	if config.SDI == 0 {
+		config.SDI = SPI0_SDI_PIN
+	}
+
+	// Default all port configuration bits to 0 for simplicity
+	avr.SPCR.Set(0)
+	avr.SPSR.Set(0)
+
+	// Setup pins output configuration
+	config.SCK.Configure(PinConfig{Mode: PinOutput})
+	config.SDI.Configure(PinConfig{Mode: PinInput})
+	config.SDO.Configure(PinConfig{Mode: PinOutput})
```

Comment:
Same here.

Reply:
Something needs to configure the pins for output mode.  If there is only one & the user asks to use the SPI port, it needs to be setup right or the SPI hardware doesn't work correctly.  I think this should stay with the configuration code.

Reply:
Yes, but it should use the fixed pins for this chip:

```suggestion
	SPI0_SCK_PIN.Configure(PinConfig{Mode: PinOutput})
	SPI0_SDI_PIN.Configure(PinConfig{Mode: PinInput})
	SPI0_SDO_PIN.Configure(PinConfig{Mode: PinOutput})
```